launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java

```diff
@@ -362,21 +363,35 @@ Map<String, String> getEffectiveConfig() throws IOException {
   }
 
   /**
-   * Loads the configuration file for the application, if it exists. This is either the
-   * user-specified properties file, or the spark-defaults.conf file under the Spark configuration
-   * directory.
+   * Load the configuration file(s) for the application - from the user-specified properties
+   * file, and/or the spark-defaults.conf file under the Spark configuration directory, if exists.
+   * Configurations from user-specified properties file take precedence over spark-defaults.conf.
    */
   private Properties loadPropertiesFile() throws IOException {
+    Properties defaultsProps = new Properties();
+    if (propertiesFile == null || loadSparkDefaults) {
+      defaultsProps = loadPropertiesFile(new File(getConfDir(), DEFAULT_PROPERTIES_FILE));
+    }
+
     Properties props = new Properties();
-    File propsFile;
     if (propertiesFile != null) {
-      propsFile = new File(propertiesFile);
+      File propsFile = new File(propertiesFile);
       checkArgument(propsFile.isFile(), "Invalid properties file '%s'.", propertiesFile);
-    } else {
-      propsFile = new File(getConfDir(), DEFAULT_PROPERTIES_FILE);
+      props = loadPropertiesFile(propsFile);
+    }
+
+    for (Map.Entry<Object, Object> entry : props.entrySet()) {
+      if (!defaultsProps.containsKey(entry.getKey())) {
+        defaultsProps.put(entry.getKey(), entry.getValue());
+      }
     }
 
-    if (propsFile.isFile()) {
+    return defaultsProps;
```

Comment:
nit: it is a bit confusing that we are returning the `defaultProps` here - maybe instead we can update the above `for` loop to iterate through `defaultsProps` and return the `props` instead.

Reply:
this mainly follows the SparkSubmit logic to merge the properties.
https://github.com/apache/spark/blob/76d2878b61bd0be878f30aff230b01cd0d5c96cb/core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala#L115-L120

for the naming concern, how about renaming `defaultProps` -> `finalProps` or `loadedProps` (I don't want to use the word "effective" since it has a different meaning in the context)

Reply:
+1 it is confusing at first glance on returning `defaultsProps`.

Reply:
sorry, the logic is wrong here, naming issue also gone after fix the logic